[FunctionComparator] Differentiate instructions passing different MDStrings

NuriAmari opened the pull request:
Prior to this patch, differing metadata operands to two otherwise identical instructions was not enough to consider the instructions different in the eyes of the function comparator. This breaks LLVM virtual function elimination, among other features.

In this patch, we handle the case where two associated operands are MDStrings of different value. This patch does not differentiate more complex metadata operands.

Commit:
[FunctionComparator] Distinguish between instructions passing different MDStrings

Prior to this patch, differing metadata operands to two otherwise
identical instructions was not enough to consider the instructions
different in the eyes of the function comparator.

In this patch, we handle the case where two associated operands are
MDStrings of different value.

This patch does not differentiate more complex metadata operands.

NuriAmari commented:
There are a couple related patches: https://github.com/llvm/llvm-project/pull/65963, https://github.com/llvm/llvm-project/commit/8325d46a4ab803c6eeff28bf1bdafb43287ce557, but they don't solve this problem.

drodriguez commented on llvm/test/Transforms/MergeFunc/mergefunc-preserve-vfe-intrinsics.ll:
```diff
@@ -0,0 +1,34 @@
+; This test contains three identical functions, aside from the metadata
+; they pass to a function call. This test verifies that the function merger
+; pass is able to merge the two functions that are truly identifical,
```
Nit: typo `identifical` → `identical`

nikic commented on llvm/test/Transforms/MergeFunc/mergefunc-preserve-vfe-intrinsics.ll:
```diff
@@ -0,0 +1,34 @@
+; pass is able to merge the two functions that are truly identifical,
+; but the third that passes different metadata is preserved
+
+; RUN: opt -passes=mergefunc -S %s | FileCheck %s
```
Use update_test_checks.py please, and move the test description below the RUN line.

nikic reviewed (approved):
LGTM

dexonsmith commented:
Can you change the subject of the commit / pull request before pushing to avoid using the phrase "distinct MDString"? I suggest, "different" instead of "distinct". The problem is that "distinct" means something specific in metadata: non-uniqued. Although this cannot be applied to `MDString`, it's still a bit confusing.

dexonsmith commented on llvm/lib/Transforms/Utils/FunctionComparator.cpp:
```diff
@@ -160,10 +160,23 @@ int FunctionComparator::cmpAttrs(const AttributeList L,
+  if (MDStringL && MDStringR) {
+    if (MDStringL == MDStringR)
+      return 0;
+    return MDStringL->getString().compare(MDStringR->getString());
```
Oh, sorry, I misread this, never mind. LG with the subject change I already requested. Give me a moment to figure out how to update the PR....

dexonsmith reviewed (approved):
LGTM, if you rephrase the subject of the commit and pull request to use "different" instead of "distinct" when referring to `MDString` (because of the language clash against distinct vs. uniqued in LLVM metadata).

Commit:
Address PR Feedback

- Use update_test_checks.py to generate the file check lines
- Move the test description below the run lines
- Fix type identifical -> identical

NuriAmari commented:
Thanks for the reviews all!
